cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -0,0 +1,270 @@
+import cpp
+import semmle.code.cpp.models.interfaces.Taint
+import semmle.code.cpp.models.interfaces.DataFlow
+
+class IteratorTraits extends Class {
+  IteratorTraits() {
+    this.hasQualifiedName("std", "iterator_traits") and
+    not this instanceof TemplateClass and
+    exists(TypedefType t |
+      this.getAMember() = t and
+      t.getName() = "iterator_category"
+    )
+  }
+
+  Type getIteratorType() { result = this.getTemplateArgument(0) }
+}
+
+class IteratorByTypedefs extends Class {
+  IteratorByTypedefs() {
+    this.getAMember().(TypedefType).hasName("difference_type") and
+    this.getAMember().(TypedefType).hasName("value_type") and
+    this.getAMember().(TypedefType).hasName("pointer") and
+    this.getAMember().(TypedefType).hasName("reference") and
+    this.getAMember().(TypedefType).hasName("iterator_category") and
+    not this.hasQualifiedName("std", "iterator_traits")
+  }
+}
+
+class StdIterator extends Class {
+  StdIterator() {
+    this.hasQualifiedName("std", "iterator")
+  }
+}
+
+class LegacyIterator extends Type {
+  LegacyIterator() {
+    this instanceof IteratorByTypedefs or
+    exists(IteratorTraits it | it.getIteratorType() = this) or
+    this instanceof StdIterator
```

Comment:
I think `LegacyIterator` could do with a better name, as it appears to include all the things we can deduce to be iterators.

Reply:
C++20 has a concept for iterators, which is supposed to replace the older definitions (`LegacyInputIterator` and the like). Since we don't have support yet, all the things we can deduce to be iterators are `LegacyIterator`s. We can call this `Iterator` for now, and break it out to cover legacy vs concept iterators if we need that distinction in the future.

Reply:
Happy with `Iterator`.